src/HexagonOptimize.cpp

```diff
@@ -333,12 +396,39 @@ Expr apply_commutative_patterns(const T *op, const vector<Pattern> &patterns, co
 
 typedef pair<Expr, Expr> MulExpr;
 
-// If ty is scalar, and x is a vector, try to remove a broadcast
-// from x prior to using lossless_cast on it.
+// If ty is scalar or a vector with different lanes count,
+// and x is a vector, try to remove a broadcast or adjust
+// the number of lanes in Broadcast or indices in a Shuffle
+// to match the ty lanes before using lossless_cast on it.
 Expr unbroadcast_lossless_cast(Type ty, Expr x) {
-    if (ty.lanes() == 1 && x.type().lanes() > 1) {
+    if (x.type().is_vector()) {
         if (const Broadcast *bc = x.as<Broadcast>()) {
-            x = bc->value;
+            if (ty.is_scalar()) {
+                x = bc->value;
+            } else {
+                x = Broadcast::make(bc->value, ty.lanes());
+            }
+        }
+        // Check if shuffle can be treated as a broadcast.
+        // For Eg: for inputs:
+        // ty -> uint32
+        // x  -> uint8 shuffle with indices:
+        //       0, 1, 2, 3, 0, 1, 2, 3, ....., 0, 1, 2, 3
+        // Here, x can be reinterpreted as a broadcast of uint32.
+        if (const Shuffle *shuff = x.as<Shuffle>()) {
+            int lanes = shuff->type.lanes();
+            int rep = ty.lanes();
```

Comment:
I think this would be better as a small helper function `is_broadcast(int src_lanes, ...)` (and maybe that should be a helper to complement https://github.com/halide/Halide/blob/master/src/IR.h#L773-L775).